x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlConfigMigrator.java

```diff
@@ -0,0 +1,386 @@
+ *
+ * There are 3 steps to the migration process
+ * 1. Read config from the clusterstate
+ *     - If a job or datafeed is added after this call it will be added to the index
+ *     - If deleted then it's possible the config will be copied before it is deleted.
+ *       Mitigate against this by filtering out jobs marked as deleting
+ * 2. Copy the config to the index
+ *     - The index operation could fail, don't delete from clusterstate in this case
+ * 3. Remove config from the clusterstate
+ *     - Before this happens config is duplicated in index and clusterstate, all ops
+ *       must prefer to use the index config at this stage
+ *     - If the clusterstate update fails then the config will remain duplicated
+ *       and the migration process should try again
+ *
+ * If there was an error in step 3 and the config is in both the clusterstate and
+ * index then when the migrator retries it must not overwrite an existing job config
+ * document as once the index document is present all update operations will function
+ * on that rather than the clusterstate
+ */
+public class MlConfigMigrator {
+
+    private static final Logger logger = LogManager.getLogger(MlConfigMigrator.class);
+
+    public static final String MIGRATED_FROM_VERSION = "migrated from version";
+
+    private final Client client;
+    private final ClusterService clusterService;
+
+    private final AtomicBoolean migrationInProgress;
+
+    public MlConfigMigrator(Client client, ClusterService clusterService) {
+        this.client = client;
+        this.clusterService = clusterService;
+        this.migrationInProgress = new AtomicBoolean(false);
+    }
+
+    /**
+     * Migrate ml job and datafeed configurations from the clusterstate
+     * to index documents.
+     *
+     * Configs to be migrated are read from the cluster state then bulk
+     * indexed into .ml-config. Those successfully indexed are then removed
+     * from the clusterstate.
+     *
+     * Migrated jobs have the job version set to v6.6.0 and the custom settings
+     * map has an entry added recording the fact the job was migrated and its
+     * original version e.g.
+     *     "migrated from version" : v6.1.0
+     *
+     *
+     * @param clusterState The current clusterstate
+     * @param listener     The success listener
+     */
+    public void migrateConfigsWithoutTasks(ClusterState clusterState, ActionListener<Boolean> listener) {
+
+        if (migrationInProgress.compareAndSet(false, true) == false) {
+            listener.onResponse(Boolean.FALSE);
+            return;
+        }
+
+        Collection<DatafeedConfig> datafeedsToMigrate = stoppedDatafeedConfigs(clusterState);
+        List<Job> jobsToMigrate = nonDeletingJobs(closedJobConfigs(clusterState)).stream()
+                .map(MlConfigMigrator::updateJobForMigration)
+                .collect(Collectors.toList());
+
+        ActionListener<Boolean> onFailureUnMarkMigrationInProgress = ActionListener.wrap(
```

Comment:
Could the variable name be just `unMarkMigrationInProgress`?  It's not just for failures.